Improve build system

teonnik opened the pull request:
I've made quite a few changes and extensions to the build script. I implemented an install target which provides an easy way of integrating the project in other CMake projects. Most of the changes are inspired by [Daniel Pfeifer's talk on Effective CMake][1]. I have tried to document what I have done in my commit messages. There are certain issues with some of the tests that still need to be resolved though. 


[1]: https://www.google.com/url?sa=t&rct=j&q=&esrc=s&source=web&cd=1&cad=rja&uact=8&ved=0ahUKEwi21Y3q673XAhXHZCYKHWXpCAUQtwIIKDAA&url=https%3A%2F%2Fwww.youtube.com%2Fwatch%3Fv%3DbsXLMQ6WgIk&usg=AOvVaw0RtjvfdZFL46eWd5uwzsmw

FelixPetriconi commented:
Thanks for helping us here!
Some questions: 
* several commits are marked as wip. Is there more progress on your side happening?
* Why did you split up the future and channel tests into more individual tests? I would prefer that at least the future tests stay in one test executable and for channels the same. Otherwise is developing on Windows a pain because of so many different projects. As well it is necessary that the helper files are used by the tests.

teonnik commented:
> several commits are marked as wip. Is there more progress on your side happening?

I have finished work on all `[wip]`. However, there were still some compilation problem with a few of the targets on my machine. (e.g. `stlab.test.channel.process`) The majority of targets compile and execute without issues. 

> Why did you split up the future and channel tests into more individual tests? I would prefer that at least the future tests stay in one test executable and for channels the same. Otherwise is developing on Windows a pain because of so many different projects.

I can group the future and channel targets. In CMake this is usually done using a `add_custom_targets`  and `add_dependencies`. It is relatively easy to modify.

> As well it is necessary that the helper files are used by the tests

This is probably what's causing the compilation problems. Although I don't understand why, the helper files are in the same directory as the files using them...

FelixPetriconi commented:
But the helper files are .cpp as well, so they have to be compiled and linked.

teonnik commented:
> But the helper files are .cpp as well, so they have to be compiled and linked.

I tried to arrange tests in similar manner to before. There are still some linkage issues. I will have to get back at it later.

teonnik commented:
The linkage errors occur with the original code as well. Either something is wrong with my current environment or the original CMake code / source code was broken. If it is the latter, then the issue is not due to changes that I made. I am not familiar enough with `boost unit test framework` to debug the issue effectively. If no one else can reproduce the linkage error, then I would know that there is something wrong on my side.

The core of the error message:

```shell
... multiple definition of `init_unit_test_suite(int, char**)'
```

FelixPetriconi commented:
I have not looked into the Mac or Linux build, but the preprocessor defines on Windows are different created by CMake on your develop branch and the stlab develop branch. This causes as far as I can see the main problem here.

teonnik commented:
When I add `#define BOOST_TEST_MODULE stlab_libraries_tests` to `main.cpp` I still get the error below. In my `develop` branch I pass the define through CMake which is equivalent. Is that the preprocessor macro that you were referring to or have I misunderstood? 

```shell
Scanning dependencies of target stlab.test.tuple
[ 33%] Building CXX object test/CMakeFiles/stlab.test.tuple.dir/tuple_test.cpp.o
[ 66%] Building CXX object test/CMakeFiles/stlab.test.tuple.dir/main.cpp.o
[100%] Linking CXX executable bin/stlab.test.tuple
CMakeFiles/stlab.test.tuple.dir/main.cpp.o: In function `init_unit_test_suite(int, char**)':
main.cpp:(.text+0x0): multiple definition of `init_unit_test_suite(int, char**)'
CMakeFiles/stlab.test.tuple.dir/tuple_test.cpp.o:tuple_test.cpp:(.text+0x1ca): first defined here
CMakeFiles/stlab.test.tuple.dir/tuple_test.cpp.o: In function `boost::unit_test::make_test_case(boost::function<void ()> const&, boost::unit_test::basic_cstring<char const>, boost::unit_test::basic_cstring<char const>, unsigned long)':
tuple_test.cpp:(.text._ZN5boost9unit_test14make_test_caseERKNS_8functionIFvvEEENS0_13basic_cstringIKcEES8_m[_ZN5boost9unit_test14make_test_caseERKNS_8functionIFvvEEENS0_13basic_cstringIKcEES8_m]+0x5d): undefined reference to `boost::unit_test::ut_detail::normalize_test_case_name[abi:cxx11](boost::unit_test::basic_cstring<char const>)'
collect2: error: ld returned 1 exit status 
test/CMakeFiles/stlab.test.tuple.dir/build.make:121: recipe for target 'test/bin/stlab.test.tuple' failed
make[3]: *** [test/bin/stlab.test.tuple] Error 1
CMakeFiles/Makefile2:169: recipe for target 'test/CMakeFiles/stlab.test.tuple.dir/all' failed
make[2]: *** [test/CMakeFiles/stlab.test.tuple.dir/all] Error 2
CMakeFiles/Makefile2:181: recipe for target 'test/CMakeFiles/stlab.test.tuple.dir/rule' failed
make[1]: *** [test/CMakeFiles/stlab.test.tuple.dir/rule] Error 2
Makefile:201: recipe for target 'stlab.test.tuple' failed
make: *** [stlab.test.tuple] Error 2
```

FelixPetriconi commented:
At least there is a define for NOMINMAX that is not resolved correctly on the Windows platform which results in a huge amount of compiler errors.

Can you build the current stlab/develop correctly without any problem? If no, then I assume something in your general setup is different to Travis, AppVeyour and our own systems.

teonnik commented:
> Can you build the current stlab/develop correctly without any problem? If no, then I assume something in your general setup is different to Travis, AppVeyour and our own systems.

I can build the `master` branch without problems. However, I get the same errors on the `develop` branch as I do on my own fork. I am using the same setup when building the `master` and the `develop` branches.

teonnik commented:
Just finished updating my fork. It's been a while since I fiddled around with the project. I will try to summarize what I have done.

I moved the `stlab` folder inside the `include` folder. It was easier to create an install target with that project structure. An install target provides an easy way to incorporate `stlab` into other `cmake` projects using `find_package(stlab)`. module files are generated automatically and can be found in `<install_directory>/lib/cmake/stlab`. To find the library,`cmake` searches system paths for the generated `.cmake` (e.g. `stlabConfig.cmake`) files. If the library is installed in a custom directory, the `CMAKE_PREFIX_PATH` library has to be set to root (`<install_directory>`). Once the library is found, an imported target `stlab::stlab` can be used easily with `target_link_libraries` by clients. Tests are run when `make install` is called.

I refactored the test infrastructure quite a bit. All tests compile and run correctly. I decoupled the `stlab.test.future` target into individual targets. That resulted in rather significant speed-up in compilation on multi-core machines. Since `Boost` is only used for unit testing, I include it in the `test/CMakeLists.txt` file only. Users of the library don't need to link against Boost if they don't plan to build the tests. Similarly, `cmake` code related to `conan` was moved to `test/CMakeLists.txt`.

There is, perhaps, a little bit more work left on correctly passing compiler flags. I'm looking forward to feedback. :)

FelixPetriconi commented:
This PR is obsolete after PR #180
